File: leet-code/297. Serialize and Deserialize Binary Tree/SerializeAndDeserializeBinaryTree297.java
package com.example;

import java.util.ArrayDeque;
import java.util.Deque;

public class SerializeAndDeserializeBinaryTree297 {

    public String serialize(TreeNode root) {
        if (root == null) return "";
        StringBuilder sb = new StringBuilder();
        sb.append(root.val).append(",");
        Deque<TreeNode> deq = new ArrayDeque<>();
        deq.add(root);

        while (!deq.isEmpty()) {
            TreeNode node = deq.poll();
            if (node.left != null) {
                sb.append(node.left.val).append(",");
                deq.add(node.left);
            } else {
                sb.append("#").append(",");
            }

            if (node.right != null) {
                sb.append(node.right.val).append(",");
                deq.add(node.right);
            } else {
                sb.append("#").append(",");
            }
        }

        return sb.toString();
    }

    public TreeNode deserialize(String data) {
        if (data.isEmpty()) return null;
        Deque<TreeNode> deq = new ArrayDeque<>();
        String[] nodeValues = data.split(",");
        TreeNode root = new TreeNode(Integer.parseInt(nodeValues[0]));
        deq.add(root);
        int idx = 0;

        while (!deq.isEmpty()) {
            TreeNode node = deq.poll();
            idx += 1;
            if (!nodeValues[idx].equals("#")) {
                node.left = new TreeNode(Integer.parseInt(nodeValues[idx]));
                deq.add(node.left);
            }

            idx += 1;
            if (!nodeValues[idx].equals("#")) {
                node.right = new TreeNode(Integer.parseInt(nodeValues[idx]));
                deq.add(node.right);
            }
        }
        return root;
    }
}

File: leet-code/297. Serialize and Deserialize Binary Tree/SerializeAndDeserializeBinaryTree297.kt
package com.example.algorithm

class SerializeAndDeserializeBinaryTree297 {
    // Encodes a URL to a shortened URL.
    fun serialize(root: TreeNode?): String {
        if (root == null) {
            return "null,"
        }

        var serialized = ""
        val queue = ArrayDeque<TreeNode?>()
        queue.add(root)
        while (queue.isNotEmpty()) {
            val node = queue.removeFirst()
            if (node == null) {
                serialized += "null,"
            } else {
                serialized += node.`val`.toString() + ","
                queue.add(node.left)
                queue.add(node.right)
            }
        }
        return serialized
    }

    // Decodes your encoded data to tree.
    fun deserialize(data: String): TreeNode? {
        if (data.startsWith("null,")) {
            return null
        }

        val words = data.split(",").filter{ it.isNotEmpty() }
        val root = TreeNode(words[0].toInt())
        val nodeQueue = ArrayDeque<TreeNode>()
        nodeQueue.add(root)

        var idx = 0
        while (nodeQueue.isNotEmpty()) {
            val node = nodeQueue.removeFirst()

            idx += 1
            if (words[idx] != "null") {
                node.left = TreeNode(words[idx].toInt())
                nodeQueue.add(node.left)
            }

            idx += 1
            if (words[idx] != "null") {
                node.right = TreeNode(words[idx].toInt())
                nodeQueue.add(node.right)
            }
        }
        return root
    }
}
